### PR TITLE
Test host tool

### PR DESCRIPTION
**Pull request by mzgoddard**
Add tests for host tool and utilities. New tests are put in `tests` directories in the directory containing the code they test. Existing `src/shared/iterate-emitter.js` tests were in such a folder looking forward to this change. That was done to help organize fixtures and snapshots. If the test files are in a `tests` directory `ava` puts snapshots by default in the `snapshots` directory under `tests`. If test files are adjacent to the code the test following the `{file}.test.js` pattern, snapshots are placed next to that with the patterns `{file}.test.js.md` and `{file}.test.js.snap`.

New tests for:
- src/host/agent.js
- src/host/cli-run-plan.js
- src/host/messages.js
- src/host/plan-form.js
- src/host/plan-object.js
- src/shared/array-util.js
- src/shared/file-glob.js
- src/shared/file-record.js
- src/shared/job.js
- src/shared/messages.js

### CODE REVIEW

**Comment by jugglinmike** [comment]
@mzgoddard Can you let me know when this is ready for review? Converting it to a draft pull request might be the way to go, though I don't know if that prevents CI from running...

**Comment by mzgoddard** [comment]
@jugglinmike This PR is ready for review.

File: src/shared/tests/file-record.js [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+  t.plan(10);
+  function assertImplementsHost(actualHost) {
+    t.is(typeof actualHost.read, 'function');
+    t.is(typeof actualHost.collapse, 'function');
+  }
+  assertImplementsHost(fileRecord.host);
+  assertImplementsHost(fileRecord.createHost());
+  assertImplementsHost(fileRecord.createHost({ fs: fs.promises, path: path }));
+  assertImplementsHost(fileRecord.createHost({ fs: { async readdir() {}, async readFile() {} } }));
+  assertImplementsHost(fileRecord.createHost({ path: { sep: '/', join() {} }, path: path.posix }));
```
Comment (jugglinmike) [inline src/shared/tests/file-record.js:47]: This object initializer has a duplicated property name. Was this meant to be two different tests?

File: src/shared/tests/file-record.js [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+  const posixHost = fileRecord.createHost({ path: path.posix });
+  t.deepEqual(posixHost.collapse({ bufferData: new Uint8Array(16) }), [
+    { bufferData: new Uint8Array(16) },
+  ]);
+  t.deepEqual(posixHost.collapse({ name: 'file', bufferData: new Uint8Array(16) }), [
+    { name: 'file', bufferData: new Uint8Array(16) },
+  ]);
+  t.deepEqual(posixHost.collapse({ entries: [] }), []);
+  t.deepEqual(posixHost.collapse({ entries: [{ name: 'file', bufferData: new Uint8Array(16) }] }), [
+    { name: 'file', bufferData: new Uint8Array(16) },
```
Comment (jugglinmike) [inline src/shared/tests/file-record.js:72]: What do you think about verifying the buffers' contents in addition to their lengths?

File: src/shared/tests/messages.js [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+  t.is(typeof logger.log, 'function');
+  t.is(typeof logger.emitter, 'object');
+  t.is(typeof logger.emitter.on, 'function');
+});
+
+test('log', async t => {
+  const { log, emitter } = messages.createSharedLogger(MESSAGES);
+  emitter.on('message', ({ text, data: { date, ...otherData } }) => {
+    // The date value is not deterministic. Check only that it is a Date.
+    t.true(date instanceof Date);
```
Comment (jugglinmike) [inline src/shared/tests/messages.js:28]: How about

```js
test('log', async t => {
  const start = new Date();
  // ...
    t.true(date >= start);
    t.true(date <= Date.now());
```

File: src/host/tests/agent.js [hunk 4]
```diff
@@ -0,0 +1,185 @@
+/// <reference path="../../data/types.js" />
+
+import test from 'ava';
+
+import { iterateEmitter } from '../../shared/iterate-emitter.js';
+import { startJob } from '../../shared/job.js';
```
Comment (jugglinmike) [inline src/host/tests/agent.js:6]: It's a little unorthodox to use the application logic to express the tests. In this context, the correctness of utilities like `iterateEmitter` and `startJob` is under scrutiny, making it risky to depend on them. If someone were modifying its behavior, then failures here would be especially difficult to interpret. They'd have to determine whether they'd invalidated the application or the test.

**Review by jugglinmike — CHANGES_REQUESTED** [review]
Glad to see this code getting tested!

File: src/shared/tests/file-record.js (hunk 2 above)
Comment (mzgoddard) [reply src/shared/tests/file-record.js:72]: Does `deepEqual` not compare typed array content? Or do you mean something like have non-zero buffer data?

File: src/host/tests/cli-run-plan.js [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    stdout: replaceNonDeterministicOutput(stdout.toString()),
+    stderr: replaceNonDeterministicOutput(stderr.toString()),
+  };
+}
+
+const ISO_DATE = /\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d+Z/g;
+const DETERMINISTIC_DATE = '2000-01-01T12:00:00.000Z';
+const URL_PATHNAME_ONLY = /'\/[^/']+((?:\/[^/']*)*')/g;
+const DETERMINISTIC_PATHNAME_PREFIX = (...match) => `'/static${match[1]}`;
+const URL_PORT_PATHNAME = /(https?:\/\/[^:]+:)\d+(\/[^/]+)?/g;
```
[the remark this answers is not in the document]
Comment (mzgoddard) [reply src/host/tests/cli-run-plan.js:69]: Thanks for noticing this. Makes a big difference.

```
  Difference:

    {
      stderr: `Starting...␊
      Reference server listening on 'http://localhost:8888'.␊
      Plan 'unknown' with 2 tests and 4 files read from source 'fork'.␊
  -   Reference available on 'http://localhost:8888/static␊
  -   Starting test agent.␊
  -   [Agent]: Starting...␊
  -   Agent running with protocol 'fork'.␊
  -   Starting test.␊
  -   [Server]: Serving '/static/reference/index.html'.␊
  -   [Agent]: Open page: 'http://localhost:8888/static/reference/index.html'.␊
  -   Starting test.␊
  -   [Server]: Serving '/static/reference/index.html'.␊
  +   Reference available on 'http://localhost:8888/static/static/reference/index.html'.␊
  +   [Agent]: Open page: 'http://localhost:8888/static/reference/index.html'.␊
  +   Starting test.␊
  +   [Server]: Serving '/static/reference/index.html'.␊
      [Agent]: Open page: 'http://localhost:8888/static/reference/index.html'.␊
  -   Removing reference from 'http://localhost:8888/static␊
  +   Removing reference from 'http://localhost:8888/static`,
```

File: src/host/tests/plan-from.js [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+    const workingdir = path.join(dirname, _workingdir);
+    for (const files of [
+      ['*.json'],
+      ['*.json', 'reference/**'],
+      ['tests/*.json'],
+      ['tests/*.json', 'reference/**'],
+    ]) {
+      for (const protocol of [undefined, 'fork', 'developer']) {
+        for (const testPattern of [undefined]) {
+          for (const log of [undefined]) {
```
[the remark this answers is not in the document]
Comment (mzgoddard) [reply src/host/tests/plan-from.js:22]: I did intend to add more. I'll add them now.

File: src/host/tests/agent.js (hunk 4 above)
Comment (mzgoddard) [reply src/host/tests/agent.js:6]: Hmm. This sounds like if that shared code was versioned and someone maintaining the harness updated to a new major version, introducing a backwards incompatible change. They would need to make that determination as well.

Should I duplicate the relevant mechanics in `iterateEmitter` and `startJob` into this file?

I think the best solution would be to change the patterns for managing this. Though that seems like a large change and I'm not sure what patterns I'd replace them with.

Am I thinking about this in a good direction?

File: src/shared/tests/file-record.js (hunk 1 above)
Comment (mzgoddard) [reply src/shared/tests/file-record.js]: Un. I meant to only have the first mirroring the mock fs createHost before it.

File: src/shared/tests/array-util.js [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    value2,
+    value3,
+  ]);
+  t.deepEqual(arrayUtil.replace([value, value2, oldValue, value3], oldValue, newValue), [
+    value,
+    value2,
+    newValue,
+    value3,
+  ]);
+  t.deepEqual(arrayUtil.replace([value, value2, value3, oldValue], oldValue, newValue), [
```
[the remark this answers is not in the document]
Comment (mzgoddard) [reply src/shared/tests/array-util.js]: Added a specific assert for this test.

**Comment by mzgoddard** [comment]
I made changes to the PR @jugglinmike. Let me know what I can do next.

File: src/host/tests/plan-from.js [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+    }),
+  };
+}
+
+/**
+ * Normalize differences in files between operating system environments.
+ * @param {AriaATCIHost.TestPlan} testPlan
+ * @returns {AriaATCIHost.TestPlan}
+ */
+function normalizeTestPlanFiles(testPlan, coders = textCoders()) {
```
[the remark this answers is not in the document]
Comment (jugglinmike) [reply src/host/tests/plan-from.js:80]: Understood. Sticking with what you've already written sounds good to me!

**Review by jugglinmike — COMMENTED** [review]
Food for thought: I studied the "cancellable job" abstraction when we introduced it a couple months ago, but I lost that context in the time since. I had to revisit the implementation to understand this patch. Do you have any trouble remembering how it works? Do you use a similar abstraction in other projects?

It makes me wonder if the "observable" pattern would have been a stronger choice here, if indeed it addresses the same issues. It's also non-standard, but it's popular among functional programming enthusiasts, even across languages. The uninitiated would still have to learn something (there's no getting around that), but the knowledge would at least be transferable.

**Pull request merged**